File: tasks/money-rollup/python/environment/src/money.py
"""Amount parsing, rounding, and formatting helpers."""

import math


def parse_amount(text, minor_units):
    return float(text)


def parse_rate(text):
    return float(text)


def round_amount(value, places):
    scale = 10**places
    return int(math.trunc(value * scale + math.copysign(0.5, value)))


def format_minor(minor, places):
    negative = minor < 0
    digits = str(-minor if negative else minor).rjust(places + 1, "0")
    if places:
        digits = digits[: len(digits) - places] + "." + digits[len(digits) - places :]
    return "-" + digits if negative else digits

File: tasks/money-rollup/javascript/environment/src/money.js
"use strict";

function parseAmount(text, minorUnits) {
  return Number(text);
}

function parseRate(text) {
  return Number(text);
}

function roundAmount(value, places) {
  const scale = 10 ** places;
  return BigInt(Math.trunc(value * scale + Math.sign(value) * 0.5));
}

function formatMinor(minor, places) {
  const negative = minor < 0n;
  let digits = (negative ? -minor : minor).toString().padStart(places + 1, "0");
  if (places) {
    const split = digits.length - places;
    digits = digits.slice(0, split) + "." + digits.slice(split);
  }
  return negative ? "-" + digits : digits;
}

module.exports = { formatMinor, parseAmount, parseRate, roundAmount };
